10/alpine/Dockerfile

```diff
@@ -37,14 +42,27 @@ RUN addgroup -g 1000 node \
     && gpg --batch --decrypt --output SHASUMS256.txt SHASUMS256.txt.asc \
     && grep " node-v$NODE_VERSION.tar.xz\$" SHASUMS256.txt | sha256sum -c - \
     && tar -xf "node-v$NODE_VERSION.tar.xz" \
+    && ln -s /usr/bin/ccache /usr/local/bin/gcc \
+    && ln -s /usr/bin/ccache /usr/local/bin/g++ \
     && cd "node-v$NODE_VERSION" \
     && ./configure \
     && make -j$(getconf _NPROCESSORS_ONLN) \
     && make install \
-    && apk del .build-deps \
     && cd .. \
     && rm -Rf "node-v$NODE_VERSION" \
-    && rm "node-v$NODE_VERSION.tar.xz" SHASUMS256.txt.asc SHASUMS256.txt
+    && rm "node-v$NODE_VERSION.tar.xz" SHASUMS256.txt.asc SHASUMS256.txt \
+    && rm /usr/local/bin/gcc /usr/local/bin/g++ \
+    # update the ccache
+    && ( \
+      [ ! "$GITHUB_API_TOKEN" ] || ( \
+          cd /root/.ccache \
+          && git add . \
+          && git -c user.name="Node.js GitHub Bot" -c user.email="bot@nodejs.org"  commit -m "Updated the cache" \
+          && git push "${CCACHE_REPO//github.com/$GITHUB_API_TOKEN:x-oauth-basic@github.com}" master > /dev/null 2>&1  || echo "Unable to push the cache update" && exit 1  \
```

Comment:
Silly Docker question, but would this mean anyone pulling the image would now need to set a GH token?

Reply:
Oh, I guess that's the if above

Reply:
@LaurentGoderre what happens when someone has the API token set, but doesn't have permission to push to that cache repo, adn hasn't overriden the `CCACHE_REPO`? EX: If they set the API token for another reason, then the image build fails right?

Reply:
@nschonni these are build instructions that are not part of the resulting images so it's a non issue for anyone getting the image from the hub or the store. For their local build to fail, the would have to explicitly send that environment variable to the docker build `docker build --build-arg GITHUB_API_TOKEN` .`

Reply:
So if someone wanted to use an alternate cache, they can easily do it by doing

`docker build --build-arg CCACHE_REPO=[alternate_repo] --build-arg GITHUB_API_TOKEN=$GITHUB_API_TOKEN 8/alpine`

Reply:
:+1: thanks for clarifying

Reply:
It seems kind of hanky to have this bit embedded directly like this in the image `Dockerfile` itself when Travis is the only place this is expected to ever be run/used. :disappointed: (Not sure a better solution though -- it really would be nice if there were official binaries published so this would be a moot point. :scream:)

As implemented, this is a bit racy, isn't it?  Doesn't Travis compile multiple Node.js versions simultaneously? (thus the possibility of one of those builds failing to `git push` because it was racing with another -- using separate Git branch for arch+major combination could combat that, but makes the repo even larger)

Reply:
We don't really use any specific Node.js `./configure` flags, so maybe just duplicating the build in a separate "pushing" `Dockerfile` would be sufficient and simple enough to maintain?

In fact, everything here is already templatized, so it could even use the same exact template and just replace a special placeholder in the real `Dockerfile` with nothing and in the "cache-generating" `Dockerfile` it gets replaced by code which pushes the results? (Then you could have Travis test both in series -- do the cache-push one first, then do the normal version; you could even use GNU `timeout` to time-box the second build to verify that it's actually getting the benefits of the caching properly.)